src/main/java/bitflow4j/steps/fork/distribute/AbstractTagFork.java

```diff
@@ -0,0 +1,129 @@
+        this.subPipelineBuilders = subPipelines;
+        availableKeys = subPipelines.stream().map(Pair::getLeft).sorted().collect(Collectors.toList());
+        compilePatterns();
+        formattedSubPipelines = ScriptableDistributor.formattedSubPipelines(subPipelineBuilders);
+    }
+
+    private void compilePatterns() throws IOException {
+        if (wildcardMatch || regexMatch) {
+            for (Pair<String, PipelineBuilder> available : subPipelineBuilders) {
+                String key = available.getLeft();
+                try {
+                    patterns.put(key, Pattern.compile(toWildcardRegex(key)));
+                } catch (PatternSyntaxException e) {
+                    throw new IOException(e);
+                }
+
+            }
+        }
+    }
+
+    private String toWildcardRegex(String key) {
+        if (wildcardMatch) {
+            return key.replaceAll("\\*", ".*");
+        }
+        return key;
+    }
+
+    @Override
+    public String toString() {
+        return "";
+    }
+
+    @Override
+    public Collection<Object> formattedChildren() {
+        return formattedSubPipelines;
+    }
+
+    @Override
+    public Collection<Pair<String, Pipeline>> distribute(Sample sample) {
+        String value = getSampleKey(sample);
+        if (value == null) {
+            value = "";
+            //TODO How to handle samples without the tag?
+        }
+        if (tagValueCache.containsKey(value)) {
+            return tagValueCache.get(value);
+        } else {
+            List<Pair<String, Pipeline>> result = new ArrayList<>();
+            if(subPipelineBuilders != null){
+                for (Pair<String, PipelineBuilder> available : subPipelineBuilders) {
+                    if (matches(available.getLeft(), value)) {
+                        Pipeline pipeline = getPipeline(available.getLeft(), value, available.getRight());
+                        if (pipeline != null) {
+                            result.add(new Pair<>(value, pipeline));
+                        }
+                    }
+                }
+            }
+            if (result.isEmpty()) {
+                logger.warning(String.format("No sub-pipeline for value %s (available keys: %s)", value, availableKeys));
+            }
+            tagValueCache.put(value, result);
+            return result;
+        }
+    }
+
+    private Pipeline getPipeline(String key, String tagValue, PipelineBuilder builder) {
+        if (subPipelines.containsKey(tagValue)) {
+            return subPipelines.get(tagValue);
+        }
+
+        try {
+            Pipeline pipe = builder.build();
+            subPipelines.put(tagValue, pipe);
+            return pipe;
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, String.format("Failed to build sub-pipeline for template value '%s' (matching %s)", tagValue, key), e);
+            return null;
+        }
+    }
+
+    protected String getSampleKey(Sample sample){
```

Comment:
The methods getSampleKey() and matches() should be abstract, since the AbstractTagFork class is abstract.